src/MarginModels/PointSetMarginModel.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2020, Battelle Energy Alliance, LLC
+# ALL RIGHTS RESERVED
+"""
+Created on May 13 2021
+
+@author: mandd,wangc
+"""
+
+#External Modules------------------------------------------------------------------------------------
+import numpy as np
+import pandas as pd
+import copy
+#External Modules End--------------------------------------------------------------------------------
+
+#Internal Modules------------------------------------------------------------------------------------
+from utils import InputData, InputTypes
+from .MarginBase import MarginBase
+from sklearn.metrics import pairwise_distances
+#Internal Modules End--------------------------------------------------------------------------------
+
+
+class PointSetMarginModel(MarginBase):
+
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Collects input specifications for this class.
+      @ In, cls, class instance
+      @ Out, inputSpecs, InputData, specs
+    """
+    inputSpecs = super(PointSetMarginModel, cls).getInputSpecification()
+    inputSpecs.description = """ PointSet Margin Model """
+    inputSpecs.addSub(InputData.parameterInputFactory('failedDataFileID', contentType=InputTypes.InterpretedListType, descr='failed data file'))
+
+    inputSpecs.addSub(InputData.parameterInputFactory('marginID', contentType=InputTypes.InterpretedListType, descr='ID of the margin variable'))
+
+    mapping = InputData.parameterInputFactory('map', contentType=InputTypes.InterpretedListType, descr='ID of the column of the csv containing failed data')
+    mapping.addParam("var", InputTypes.StringType)
+    inputSpecs.addSub(mapping)
+
+    return inputSpecs
+
+
+  def __init__(self):
+    """
+      Constructor
+      @ In, None
+      @ Out, None
+    """
+    MarginBase.__init__(self)
+
+    self.failedDataFileID = None  # name of the file containing the failed data
+    self.mapping = {}             # dictionary containing mapping between actual and failed data
+    self.InvMapping = {}          # dictionary containing mapping between failed and actual data
+    self.marginID = None          # ID of the calculated margin variable
+    self.dimensionality = None    # dimensionality of the point set
+
+  def _handleInput(self, paramInput):
+    """
+      Function to read the portion of the parsed xml input that belongs to this specialized class
+      and initialize some stuff based on the inputs got
+      @ In, paramInput, InputData.ParameterInput, the parsed xml input
+      @ Out, None
+    """
+    super()._handleInput(paramInput)
+    for child in paramInput.subparts:
+      if child.getName() == 'failedDataFileID':
+        self.failedDataFileID = self.setVariable(child.value)
```

Comment:
I think the method "setVariable" is not used correctly. The purpose of setVariable is to deal with the situation when we want to treat the node to be either raven variables or user inputs. The right way to use it is:
```
self.variableName = self.setVariable(variableValue)
self._variableDict[variableName] = self.variableName
```

The second line is required if you want to perturb it through RAVEN. In addition, I think we need to put the second line directly into the method ```setVariables```, currently, it is easy to forget it.

Reply:
I can changed it if you want.

Reply:
I did these changes

      if child.getName() == 'failedDataFileID':
        self.failedDataFileID = self.setVariable(child.value)
        self._variableDict['failedDataFileID'] = self.failedDataFileID
      if child.getName() == 'marginID':
        self.marginID = self.setVariable(child.value)
        self._variableDict['marginID'] = self.marginID

but it gives me an error:

Traceback (most recent call last):
  File "/Users/mandd/miniconda2/envs/raven_libraries/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/Users/mandd/miniconda2/envs/raven_libraries/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/mandd/projects/raven/framework/Runners/SharedMemoryRunner.py", line 114, in <lambda>
    self.thread = InterruptibleThread(target = lambda q, *arg : q.append(self.functionToRun(*arg)),
  File "/Users/mandd/projects/raven/framework/Models/ExternalModel.py", line 282, in evaluateSample
    result,instSelf = self._externalRun(inRun,Input[1],) #entry [1] is the external model object; it doesn't appear to be needed
  File "/Users/mandd/projects/raven/framework/Models/ExternalModel.py", line 229, in _externalRun
    self.sim.run(externalSelf, InputDict)
  File "/Users/mandd/projects/raven/plugins/SR2ML/src/MarginModel.py", line 78, in run
    self._model.initialize(inputDict)
  File "/Users/mandd/projects/raven/plugins/SR2ML/src/MarginModels/PointSetMarginModel.py", line 88, in initialize
    super().initialize(inputDict)
  File "/Users/mandd/projects/raven/plugins/SR2ML/src/Bases/ModelBase.py", line 81, in initialize
    needDict = self.loadVariables(needDict, inputDict)
  File "/Users/mandd/projects/raven/plugins/SR2ML/src/Bases/ModelBase.py", line 134, in loadVariables
    raise KeyError('Looking for variable "{}" to fill "{}" but not found among variables!'.format(val, key))
KeyError: 'Looking for variable "failureData.csv" to fill "failedDataFileID" but not found among variables!'

Reply:
This is because setVariable is used to parse the node value. The idea is that if you do not want to treat it as RAVEN variable, just use the InputData to parse the node.

Reply:
By the way, I have created a pull request to combine self.setVariable and self._variableDict #30